File: py/wlearn/automl/_executor.py
"""Executor matching JS automl/executor.js."""

import math

import numpy as np

from ._leaderboard import Leaderboard
from ._common import now, seed_for, partial_shuffle
from ._cv import get_scorer
from ._rng import make_lcg


class Executor:
    """Evaluation engine: evaluates candidates across CV folds, applies budgets."""

    def __init__(self, folds, scoring, X, y, time_limit_ms=0, seed=42):
        """
        Args:
            folds: list of (train_indices, test_indices) np.int32 arrays
            scoring: str or callable
            X: np.ndarray feature matrix
            y: np.ndarray labels
            time_limit_ms: global time limit (0 = no limit)
            seed: base seed for reproducibility
        """
        self._folds = folds
        self._scorer_fn = get_scorer(scoring)
        self._X = X
        self._y = y
        self._time_limit_ms = time_limit_ms
        self._seed = seed
        self._start_time = now()
        self._leaderboard = Leaderboard()

    @property
    def leaderboard(self):
        return self._leaderboard

    @property
    def is_timed_out(self):
        if self._time_limit_ms <= 0:
            return False
        return (now() - self._start_time) > self._time_limit_ms

    def evaluate_candidate(self, candidate_id, cls, params, budget=None):
        """Evaluate one candidate across all CV folds.

        Args:
            candidate_id: stable identifier
            cls: estimator class with create/fit/predict/dispose
            params: hyperparameters
            budget: optional dict with 'type' and 'value'

        Returns:
            CandidateResult dict
        """
        folds = self._folds
        scores = np.zeros(len(folds), dtype=np.float64)
        t0 = now()
        total_train_used = 0

        effective_params = self._apply_rounds_budget(cls, params, budget)

        for f, (train, test) in enumerate(folds):
            # Apply subsample budget to train only
            if budget and budget.get('type') == 'subsample':
                train = self._subsample_train(train, budget['value'], candidate_id, f)

            total_train_used += len(train)

            X_train, y_train = self._X[train], self._y[train]
            X_test, y_test = self._X[test], self._y[test]

            model = cls.create(effective_params)
            try:
                model.fit(X_train, y_train)
                preds = model.predict(X_test)
                scores[f] = self._scorer_fn(y_test, preds)
            finally:
                model.dispose()

        fit_time_ms = now() - t0

        entry = self._leaderboard.add(
            model_name=candidate_id.split(':')[0],
            params=params,
            scores=scores,
            fit_time_ms=fit_time_ms,
        )

        return {
            'candidateId': candidate_id,
            'meanScore': entry['meanScore'],
            'foldScores': scores,
            'stdScore': entry['stdScore'],
            'fitTimeMs': fit_time_ms,
            'nTrainUsed': round(total_train_used / len(folds)),
            'nTest': len(folds[0][1]),
        }

    def _apply_rounds_budget(self, cls, params, budget):
        """Apply rounds budget by setting the model's rounds param."""
        if not budget or budget.get('type') != 'rounds':
            return params
        spec = None
        if hasattr(cls, 'budget_spec'):
            spec = cls.budget_spec()
        if not spec or 'roundsParam' not in spec:
            return params
        rounds_param = spec['roundsParam']
        if rounds_param in params:
            return params
        return {**params, rounds_param: budget['value']}

    def _subsample_train(self, train, fraction, candidate_id, fold_idx):
        """Subsample train indices using partial Fisher-Yates."""
        k = max(1, math.ceil(len(train) * fraction))
        if k >= len(train):
            return train
        copy = np.array(train, dtype=np.int32)
        s = seed_for(candidate_id, fold_idx, self._seed)
        rng = make_lcg(s)
        return partial_shuffle(copy, k, rng)

    def run_strategy(self, strategy):
        """Run a strategy to completion.

        Returns:
            dict with 'leaderboard'
        """
        while not strategy.is_done():
            if self.is_timed_out:
                break
            task = strategy.next()
            if task is None:
                break
            result = self.evaluate_candidate(
                task['candidateId'],
                task['cls'],
                task['params'],
                task.get('budget'),
            )
            strategy.report(result)
        return {'leaderboard': self._leaderboard}

File: packages/automl/src/executor.js
const { normalizeX, normalizeY, makeLCG, getScorer } = require('@wlearn/core')
const { Leaderboard } = require('./leaderboard.js')
const { now, seedFor, partialShuffle } = require('./common.js')

const { ceil, min } = Math

/**
 * Subset rows of X by index array.
 */
function subsetX(X, indices) {
  const { data, cols } = X
  const rows = indices.length
  const out = new Float64Array(rows * cols)
  for (let i = 0; i < rows; i++) {
    const srcOff = indices[i] * cols
    out.set(data.subarray(srcOff, srcOff + cols), i * cols)
  }
  return { data: out, rows, cols }
}

/**
 * Subset labels by index array.
 */
function subsetY(y, indices) {
  const out = new (y.constructor)(indices.length)
  for (let i = 0; i < indices.length; i++) {
    out[i] = y[indices[i]]
  }
  return out
}

/**
 * Executor: evaluation engine and canonical leaderboard owner.
 *
 * Evaluates candidates across all CV folds, applies budgets,
 * records results in a single Leaderboard instance.
 */
class Executor {
  #folds
  #scorerFn
  #X
  #y
  #timeLimitMs
  #seed
  #startTime
  #leaderboard
  #onProgress

  /**
   * @param {object} opts
   * @param {Array<{train: Int32Array, test: Int32Array}>} opts.folds - CV folds
   * @param {string|Function} opts.scoring - scorer name or function
   * @param {object} opts.X - normalized feature matrix
   * @param {TypedArray} opts.y - normalized labels
   * @param {number} opts.timeLimitMs - global time limit (0 = no limit)
   * @param {number} opts.seed - base seed for reproducibility
   * @param {Function} opts.onProgress - optional progress callback
   */
  constructor({ folds, scoring, X, y, timeLimitMs = 0, seed = 42, onProgress }) {
    this.#folds = folds
    this.#scorerFn = getScorer(scoring)
    this.#X = X
    this.#y = y
    this.#timeLimitMs = timeLimitMs
    this.#seed = seed
    this.#startTime = now()
    this.#leaderboard = new Leaderboard()
    this.#onProgress = onProgress || null
  }

  get leaderboard() {
    return this.#leaderboard
  }

  get isTimedOut() {
    if (this.#timeLimitMs <= 0) return false
    return (now() - this.#startTime) > this.#timeLimitMs
  }

  /**
   * Evaluate one candidate across all CV folds.
   *
   * @param {object} candidateEval
   * @param {string} candidateEval.candidateId - stable identifier
   * @param {object} candidateEval.cls - estimator class with create/fit/predict/dispose
   * @param {object} candidateEval.params - hyperparameters
   * @param {object} [candidateEval.budget] - optional budget constraint
   * @returns {Promise<object>} CandidateResult
   */
  async evaluateCandidate({ candidateId, cls, params, budget }) {
    const folds = this.#folds
    const scores = new Float64Array(folds.length)
    const t0 = now()
    let totalTrainUsed = 0

    // Resolve effective params (apply rounds budget if applicable)
    const effectiveParams = this.#applyRoundsBudget(cls, params, budget)

    for (let f = 0; f < folds.length; f++) {
      let { train, test } = folds[f]

      // Apply subsample budget to train only
      if (budget && budget.type === 'subsample') {
        train = this.#subsampleTrain(train, budget.value, candidateId, f)
      }

      totalTrainUsed += train.length

      const Xtrain = subsetX(this.#X, train)
      const ytrain = subsetY(this.#y, train)
      const Xtest = subsetX(this.#X, test)
      const ytest = subsetY(this.#y, test)

      const model = await cls.create(effectiveParams)
      try {
        model.fit(Xtrain, ytrain)
        const preds = await model.predict(Xtest)
        scores[f] = this.#scorerFn(ytest, preds)
      } finally {
        model.dispose()
      }
    }

    const fitTimeMs = now() - t0

    // Record in leaderboard
    const entry = this.#leaderboard.add({
      modelName: candidateId.split(':')[0],
      params,
      scores,
      fitTimeMs,
    })

    return {
      candidateId,
      meanScore: entry.meanScore,
      foldScores: scores,
      stdScore: entry.stdScore,
      fitTimeMs,
      nTrainUsed: Math.round(totalTrainUsed / folds.length),
      nTest: folds[0].test.length,
    }
  }

  /**
   * Apply rounds budget by setting the model's rounds param if:
   * 1. Budget type is 'rounds'
   * 2. Model exposes budgetSpec().roundsParam
   * 3. Candidate params don't already set that param (candidate config wins)
   */
  #applyRoundsBudget(cls, params, budget) {
    if (!budget || budget.type !== 'rounds') return params
    const spec = cls.budgetSpec?.()
    if (!spec || !spec.roundsParam) return params
    if (params[spec.roundsParam] !== undefined) return params
    return { ...params, [spec.roundsParam]: budget.value }
  }

  /**
   * Subsample train indices using partial Fisher-Yates with deterministic seed.
   * Returns a new array of selected indices. Test indices are never subsampled.
   */
  #subsampleTrain(train, fraction, candidateId, foldIdx) {
    const k = Math.max(1, ceil(train.length * fraction))
    if (k >= train.length) return train
    // Copy to avoid mutating the original fold indices
    const copy = new Int32Array(train)
    const seed = seedFor(candidateId, foldIdx, this.#seed)
    const rng = makeLCG(seed)
    return partialShuffle(copy, k, rng)
  }

  /**
   * Run a strategy to completion.
   * Returns { leaderboard } only. Callers decide "best".
   */
  async runStrategy(strategy) {
    let done = 0
    while (!strategy.isDone()) {
      if (this.isTimedOut) break
      const task = strategy.next()
      if (task === null) break
      try {
        const result = await this.evaluateCandidate(task)
        strategy.report(result)
        done++
        if (this.#onProgress) {
          const best = this.#leaderboard.best()
          this.#onProgress({
            phase: 'search',
            candidatesDone: done,
            bestScore: best ? best.meanScore : null,
            bestModel: best ? best.modelName : null,
            lastCandidate: {
              model: result.candidateId.split(':')[0],
              score: result.meanScore,
              timeMs: result.fitTimeMs,
            },
            elapsedMs: now() - this.#startTime,
          })
        }
      } catch {
        done++
        // Skip failed candidates (invalid params, create errors, etc.)
      }
    }
    return { leaderboard: this.#leaderboard }
  }
}

module.exports = { Executor }
